tests/test_squeues.py

```diff
@@ -17,20 +18,22 @@ class TestLoader(ItemLoader):
     name_out = staticmethod(_test_procesor)
 
 def nonserializable_object_test(self):
+    q = self.queue()
     try:
         pickle.dumps(lambda x: x)
     except Exception:
         # Trigger Twisted bug #7989
         import twisted.persisted.styles  # NOQA
-        q = self.queue()
         self.assertRaises(ValueError, q.push, lambda x: x)
     else:
         # Use a different unpickleable object
         class A(object): pass
         a = A()
         a.__reduce__ = a.__reduce_ex__ = None
-        q = self.queue()
         self.assertRaises(ValueError, q.push, a)
+    # Selectors should fail (lxml.html.HtmlElement objects can't be pickled)
+    sel = Selector(text='<html><body><p>some text</p></body></html>')
```

Comment:
This line rang a bell for me. 

Based on the test case, I think it is clear that Selector instances can't be pickled so let's enforce it by adding a `__getstate__` method that enforces it (it is also quicker than going down into its `._root` element to find out it fails.
   
Aside of that, it is unclear to me if `lxml.html.HTMLElement` should be pickeable or not. I don't think https://bugs.launchpad.net/lxml/+bug/736708 is relevant.

Reply:
Interesting thread on the topic that recommends raising `TypeError` for unpickeable objects https://mail.python.org/pipermail/python-list/2014-April/671009.html

Reply:
How would you pickle HtmlElement (or other Element)? As I understand, to do this you need to save the whole tree along with an element, and this can be very inefficient if you're saving many elements.

If HtmlElement shouldn't be picklable, it should be a bug report to lxml, not a monkey patch.

It seems that picking of Selector objects is tricky indeeed; +1 to disallow pickling of Selectors explicitly, a great idea. It will be easy to re-enable it if we ever figure out how to pickle lxml sutuctures, and it'd prevent errors in the meantime.

Reply:
Thanks for your comments guys 😄 

As it says in https://bugs.launchpad.net/lxml/+bug/736708: "This has been discussed on the mailing list several times, please read the archives for reasons why this is not trivial". Seems like the archives search site is down though (or I couldn't find it :sweat_smile:)

I don't think there's anything inherently unpickable about the `Selector` class, in fact it could probably be pickled storing the text attribute instead of the `HtmlElement` object, and recreate it when unpickling (not saying it would be efficient or a good idea, just _possible_, I think).

Anyway, defining a `__getstate__` method that raises `TypeError` sounds good to me, but it should probably go in `parsel`, right? I see Scrapy's `Selector` inherits from it.

Reply:
Yeah, the fix should be in parsel.

Storing text of HtmlElement is not sufficient, as you can query siblings or parents of an element.

Reply:
should we close and implement changes in parsel then? 

anyone going to report the issue upstream to lxml?

Reply:
I think that at least tests from this PR can be useful, so we shouldn't necessarily close it. +1 to open lxml issue and fix it in parsel - @elacuesta do you have plans to do it?

Reply:
Yes @kmike, I plan to open a PR to `parsel`. I think this PR shouldn't be closed, because the change to catch `TypeError`in `scrapy/sques.py` is still necessary. I'll update this shortly.